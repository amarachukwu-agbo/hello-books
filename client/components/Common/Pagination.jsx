import React from 'react';
import propTypes from 'prop-types';

/**
 * @description common pagination component for components
 * that dispaly large data
 *
 * @param {object} pagination - pagination object
 * @param {boolean} floatingButton - determines if pagination is done
 * using only floating buttons or with buttons and numbers displaying pages
 * @param {function} onPageChange - function to be executed
 * when current page changes
 *
 * @returns {Node} - react node containing the pagination component
 */

const Pagination = ({ pagination, onPageChange, floatingButton }) => {
  const {
    currentPage,
    dataCount,
    pageCount,
  } = pagination;
  const allowPrevious = currentPage !== 1;
  const allowNext = currentPage !== pageCount;
  const previousPage = currentPage - 1;
  const nextPage = currentPage + 1;

  // Render for pagination using only floating button
  if (floatingButton) {
    return (
      <div className="row books-list">
        {
          allowPrevious &&
          <button className="btn-flat btn-small
            waves-effect waves-light white red-text left"
            onClick = {() => onPageChange(previousPage)}
          >
            <i className="fa fa-angle-double-left next-icon" />
          </button>
        }
        {
          allowNext &&
          <button className="btn-flat btn-small
            waves-effect waves-light white red-text right"
            onClick = {() => onPageChange(nextPage)}
          >
            <i className="fa fa-angle-double-right next-icon" />
          </button>
        }
      </div>
    );
  }

  // Render for pagination requiring buttons and page numbers
  return (
    <div className="pagination">
      <div className="pagination-item">
      <span>{ dataCount } Results</span>
        Displaying Page
        <span> { currentPage } </span> of
        <span> { pageCount } </span>
      </div>
      {
        allowPrevious &&
        <div className="pagination-item">
          <button className= "primary-button" onClick = { () =>
            onPageChange(previousPage)}> Previous
          </button></div>
      }
      {
        allowNext &&
        <div className="pagination-item">
          <button className= "primary-button" onClick = { () =>
            onPageChange(nextPage) }> Next
          </button></div>
      }
    </div>
  );
};

// Prop type validation
Pagination.propTypes = {
  onPageChange: propTypes.func.isRequired,
  pagination: propTypes.shape({
    currentPage: propTypes.number.isRequired,
    dataCount: propTypes.number.isRequired,
    pageSize: propTypes.number.isRequired,
    pageCount: propTypes.number.isRequired,
  }),
};

export default Pagination;
