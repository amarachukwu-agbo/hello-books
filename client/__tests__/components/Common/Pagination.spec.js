import React from 'react';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import Pagination from '../../../components/Common/Pagination';

let wrapper;

const props = {
  pagination: {
    currentPage: 4,
    dataCount: 20,
    pageCount: 8,
  },
  onPageChange: sinon.spy(() => Promise.resolve()),
};

describe('<Pagination />', () => {
  beforeEach(() => {
    wrapper = shallow(<Pagination {...props}/>);
  });
  it(`calls the onPageChange() method when
    the previous button is clicked`, () => {
    const previousButton = wrapper.find('#previous');
    previousButton.simulate('click');
    expect(props.onPageChange.called).toEqual(true);
    expect(props.onPageChange.calledWith(3)).toEqual(true);
  });

  it(`calls the onPageChange() method when
    the next button is clicked`, () => {
    const nextButton = wrapper.find('#next');
    nextButton.simulate('click');
    expect(props.onPageChange.called).toEqual(true);
    expect(props.onPageChange.calledWith(5)).toEqual(true);
  });

  it('renders correctly when floating button is true', () => {
    wrapper.setProps({
      floatingButton: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it(`calls the onPageChange() method when floating button is true and
    the previous button is clicked`, () => {
    wrapper.setProps({
      floatingButton: true,
    });
    const previousButton = wrapper.find('#floating-previous');
    previousButton.simulate('click');
    expect(props.onPageChange.called).toEqual(true);
    expect(props.onPageChange.calledWith(3)).toEqual(true);
  });

  it(`calls the onPageChange() method when floating button is true and
    the next button is clicked`, () => {
    wrapper.setProps({
      floatingButton: true,
    });
    const nextButton = wrapper.find('#floating-next');
    nextButton.simulate('click');
    expect(props.onPageChange.called).toEqual(true);
    expect(props.onPageChange.calledWith(5)).toEqual(true);
  });
});
