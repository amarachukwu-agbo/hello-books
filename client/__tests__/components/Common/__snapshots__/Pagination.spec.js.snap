// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders correctly when floating button is true 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    floatingButton={true}
    onPageChange={[Function]}
    pagination={
      Object {
        "currentPage": 4,
        "dataCount": 20,
        "pageCount": 8,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="btn-flat btn-small waves-effect waves-light white red-text left"
          id="floating-previous"
          onClick={[Function]}
        >
          <i
            className="fa fa-angle-double-left next-icon"
          />
        </button>,
        <button
          className="btn-flat btn-small waves-effect waves-light white red-text right"
          id="floating-next"
          onClick={[Function]}
        >
          <i
            className="fa fa-angle-double-right next-icon"
          />
        </button>,
      ],
      "className": "row books-list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            className="fa fa-angle-double-left next-icon"
          />,
          "className": "btn-flat btn-small waves-effect waves-light white red-text left",
          "id": "floating-previous",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fa fa-angle-double-left next-icon",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            className="fa fa-angle-double-right next-icon"
          />,
          "className": "btn-flat btn-small waves-effect waves-light white red-text right",
          "id": "floating-next",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fa fa-angle-double-right next-icon",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="btn-flat btn-small waves-effect waves-light white red-text left"
            id="floating-previous"
            onClick={[Function]}
          >
            <i
              className="fa fa-angle-double-left next-icon"
            />
          </button>,
          <button
            className="btn-flat btn-small waves-effect waves-light white red-text right"
            id="floating-next"
            onClick={[Function]}
          >
            <i
              className="fa fa-angle-double-right next-icon"
            />
          </button>,
        ],
        "className": "row books-list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="fa fa-angle-double-left next-icon"
            />,
            "className": "btn-flat btn-small waves-effect waves-light white red-text left",
            "id": "floating-previous",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-angle-double-left next-icon",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="fa fa-angle-double-right next-icon"
            />,
            "className": "btn-flat btn-small waves-effect waves-light white red-text right",
            "id": "floating-next",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-angle-double-right next-icon",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
